src/core/server/saved_objects/service/lib/collect_multi_namespace_references.ts

```diff
@@ -0,0 +1,306 @@
+  /** The space(s) that the referenced object exists in */
+  spaces: string[];
+  /**
+   * References to this object; note that this does not contain _all inbound references everywhere for this object_, it only contains
+   * inbound references for the scope of this operation
+   */
+  inboundReferences: Array<{
+    /** The type of the object that has the inbound reference */
+    type: string;
+    /** The ID of the object that has the inbound reference */
+    id: string;
+    /** The name of the inbound reference */
+    name: string;
+  }>;
+  /** Whether or not this object or reference is missing */
+  isMissing?: boolean;
+  /** The space(s) that legacy URL aliases matching this type/id exist in */
+  spacesWithMatchingAliases?: string[];
+}
+
+/**
+ * The response when object references are collected.
+ *
+ * @public
+ */
+export interface SavedObjectsCollectMultiNamespaceReferencesResponse {
+  objects: SavedObjectReferenceWithContext[];
+}
+
+export interface CollectMultiNamespaceReferencesParams {
+  registry: ISavedObjectTypeRegistry;
+  allowedTypes: string[];
+  client: RepositoryEsClient;
+  serializer: SavedObjectsSerializer;
+  getIndexForType: (type: string) => string;
+  objects: SavedObjectsCollectMultiNamespaceReferencesObject[];
+  options?: SavedObjectsCollectMultiNamespaceReferencesOptions;
+}
+
+/**
+ * Additional options exposed for unit tests.
+ *
+ * @internal
+ */
+export interface CollectMultiNamespaceReferencesParamsInternal
+  extends CollectMultiNamespaceReferencesParams {
+  /** Default: 1000 */
+  searchPerPage?: number;
+}
+
+/**
+ * Gets all references and transitive references of the given objects. Ignores any object and/or reference that is not a multi-namespace
+ * type.
+ */
+export async function collectMultiNamespaceReferences({
+  registry,
+  allowedTypes,
+  client,
+  serializer,
+  getIndexForType,
+  objects,
+  options = {},
+  searchPerPage,
+}: CollectMultiNamespaceReferencesParamsInternal): Promise<SavedObjectsCollectMultiNamespaceReferencesResponse> {
+  if (!objects.length) {
+    return { objects: [] };
+  }
+  const { namespace, purpose, typesToExclude = [] } = options;
+  const inboundReferencesMap = objects.reduce(
+    // Add the input objects to the references map so they are returned with the results, even if they have no inbound references
+    (acc, { type, id }) => acc.set(`${type}:${id}`, new Map()),
+    new Map<string, Map<string, string>>()
+  );
+  const objectMap = new Map<string, SavedObject | null>();
+
+  const rootFields = getRootFields();
+  const makeBulkGetDocs = (objectsToGet: SavedObjectsCollectMultiNamespaceReferencesObject[]) =>
+    objectsToGet.map(({ type, id }) => ({
+      _id: serializer.generateRawId(undefined, type, id),
+      _index: getIndexForType(type),
+      _source: rootFields, // Optimized to only retrieve root fields (ignoring type-specific fields)
+    }));
+  const objectFilter = ({ type }: SavedObjectsCollectMultiNamespaceReferencesObject) =>
```

Comment:
nit: it would make it easier to read the code if the name of this function reflected what it was filtering for. Then the useage below would make more sense (currently it just `objects.filter(objectFilter)`, which doesn't really mean anything).